Hhh 11147

sebersole opened the pull request:
https://hibernate.atlassian.net/browse/HHH-11147

Adds the ability to use enhanced entities as a "proxy", meaning being able to return an entity knowing just its identifier using a new interceptor implementation.

dreab8 commented on hibernate-core/src/main/java/org/hibernate/bytecode/enhance/spi/interceptor/AbstractInterceptor.java:
```diff
@@ -0,0 +1,161 @@
+/**
+ * @author Steve Ebersole
+ */
+public abstract class AbstractInterceptor implements PersistentAttributeInterceptor, SessionAssociableInterceptor {
```
just a minor, `PersistentAttributeInterceptor` can be removed, `SessionAssociableInterceptor` already extends `PersistentAttributeInterceptor`

gbadner replied:
Also, ```AbstractSessionAssociableInterceptor``` might be a better name.

gbadner replied:
Should this class implement ```Serializable```?

dreab8 commented on hibernate-core/src/main/java/org/hibernate/bytecode/enhance/spi/interceptor/BytecodeLazyAttributeInterceptor.java:
```diff
@@ -0,0 +1,41 @@
+ * @author Steve Ebersole
+ */
+@Incubating
+public interface BytecodeLazyAttributeInterceptor extends PersistentAttributeInterceptor, SessionAssociableInterceptor {
```
as in the previous comment, `PersistentAttributeInterceptor` can be removed.

dreab8 commented on hibernate-core/src/test/java/org/hibernate/test/bytecode/enhancement/lazy/group/ComplexFetchGroupTest.java:
```diff
@@ -25,80 +25,200 @@
 	@Test
-	public void testNonOwningOneToOneAccess() {
+	public void testLoadNonOwningOneToOne() {
+		// Test loading D and accessing E
```
I think we can remove this comment

dreab8 commented on hibernate-core/src/main/java/org/hibernate/event/internal/DefaultMergeEventListener.java:
```diff
@@ -191,36 +208,54 @@ protected void entityIsTransient(MergeEvent event, Map copyCache) {
 
 		final Object entity = event.getEntity();
-		final EventSource source = event.getSession();
+		if ( entity instanceof PersistentAttributeInterceptable ) {
```
this check can be removed

dreab8 commented on hibernate-core/src/main/java/org/hibernate/event/internal/DefaultLoadEventListener.java:
```diff
@@ -253,17 +253,27 @@ private Object proxyOrLoad(
-			final Object proxy = persistenceContext.getProxy( keyToLoad );
-			if ( proxy != null ) {
-				return proxy;
+			if ( persister.hasProxy() ) {
```
inside this block  `persister.getEntityMetamodel().getBytecodeEnhancementMetadata().isEnhancedForLazyLoading()` have to be `true` so `persister.hasProxy()` so if i do no miss something based on the implementation 
`public boolean hasProxy() {
		// skip proxy instantiation if entity is bytecode enhanced
		return entityMetamodel.isLazy() && !entityMetamodel.getBytecodeEnhancementMetadata().isEnhancedForLazyLoading();
	}`
it will be always `false`

sebersole replied:
* `#hasProxy` - can it produce HibernateProxy instances
* `#isEnhancedForLazyLoading` - is the entity enhanced

dreab8 commented on hibernate-core/src/main/java/org/hibernate/internal/SessionImpl.java:
```diff
@@ -1171,6 +1181,7 @@ public final Object internalLoad(
 				event.setEntityId( null );
 				event.setInstanceToLoad( null );
 				event.setResult( null );
+				event.setShouldUnwrapProxy( unwrapProxy );
```
I think the set should be moved before the `fireLoad( event, type );` at line 1174.

gbadner commented:
Rebased master with commit that skips failing OffsetTime tests, and force-pushed.

sebersole commented:
The merge tests seemed wrong to me.  I changed the tests and all 3 pass - I was kind of surprised the "merge detached uninitialized bytecode proxy" passed as-is.

Anyway, fixed tests, squashed and force pushed.

gbadner commented on hibernate-core/src/main/java/org/hibernate/persister/entity/AbstractEntityPersister.java:
```diff
@@ -4279,6 +4307,50 @@ public Object load(Serializable id, Object optionalObject, LockOptions lockOptio
+					entity,
+					proxyInterceptor.getEntityKey().getIdentifier(),
+					session
+			);
```
IIUC ```PojoEntityTuplizer#afterInitialize``` has already injected the ```LazyAttributeLoadingInterceptor```. I don't think it hurts anything, but seems redundant.

gbadner commented on hibernate-core/src/main/java/org/hibernate/Hibernate.java:
```diff
@@ -187,7 +202,10 @@ public static boolean isPropertyInitialized(Object proxy, String propertyName) {
+				return false;
+			}
+			if ( interceptor instanceof LazyAttributeLoadingInterceptor ) {
 				return ( (LazyAttributeLoadingInterceptor) interceptor ).isAttributeLoaded( propertyName );
```
Currently, ```Hibernate.isPropertyInitialized( activity, "instruction" )``` will return ```true``` if the attribute is initialized to an uninitialized proxy.

At the same time, ```Hibernate.isInitialized( activity.getInstruction() )``` will return ```false```.

I find this confusing. IMO, the return values for both should be ```false```.

gbadner replied:
I see that in ```FetchGraphTest#testLoadNonOwningOneToOne```, that this is intended.

It would be helpful to add documentation to this method that a return value of ```true``` could indicate that the property can be initialized to an uninitialized collection or proxy.

sebersole replied:
I disagree.  The contract of isPropertyInitialized is whether that value is available on the class

sebersole replied:
Unless I am off base, the same should be true in 4.x if an object is enhanced and has a proxy for an association.

gbadner commented on hibernate-core/src/main/java/org/hibernate/boot/spi/SessionFactoryOptions.java:
```diff
@@ -308,4 +308,8 @@ default boolean areJPACallbacksEnabled() {
+	/**
+	 * Can bytecode-enhanced entity classes be used as a "proxy"?
+	 */
+	boolean isEnhancementAsProxyEnabled();
```
Since this is a new SPI method, shouldn't this be added as a default method returning ```false```?

sebersole replied:
Maybe

gbadner commented on hibernate-core/src/main/java/org/hibernate/bytecode/enhance/spi/interceptor/AbstractInterceptor.java:
```diff
@@ -0,0 +1,161 @@
+		if ( session != null && !allowLoadOutsideTransaction ) {
+			this.allowLoadOutsideTransaction = session.getFactory().getSessionFactoryOptions().isInitializeLazyStateOutsideTransactionsEnabled();
+			if ( this.allowLoadOutsideTransaction ) {
+				this.sessionFactoryUuid = session.getFactory().getUuid();
```
Is it possible that this ```AbstractInterceptor``` will become associated with a ```Session``` from a different ```SessionFactory```? If so, ```sessionFactoryUuid``` won't be updated if ```allowLoadOutsideTransaction``` == ```true```.

sebersole replied:
Have you run into this before?  This is exactly the code as it happened before this work.  I see what you are saying, but I think that has nothing to do with this work.

gbadner commented on hibernate-core/src/main/java/org/hibernate/bytecode/enhance/spi/interceptor/AbstractInterceptor.java:
```diff
@@ -0,0 +1,161 @@
+
+	@Override
+	public void setSession(SharedSessionContractImplementor session) {
+		this.session = session;
```
Should further checks be done when  ```this.session != null && session != null```, similar to what is done in [```AbstractPersistentCollection#setCurrentSession```](https://github.com/hibernate/hibernate-orm/blob/master/hibernate-core/src/main/java/org/hibernate/collection/internal/AbstractPersistentCollection.java#L678-L698)?

sebersole replied:
It would not hurt I guess.  But again this was taken directly from the existing interceptor code.  Have you seen trouble with that?  I think this is another one of "not specific to this work"

gbadner commented on hibernate-core/src/main/java/org/hibernate/bytecode/enhance/spi/interceptor/EnhancementAsProxyLazinessInterceptor.java:
```diff
@@ -0,0 +1,165 @@
+		this.nonAggregatedCidMapper = nonAggregatedCidMapper;
+		assert nonAggregatedCidMapper != null || identifierAttributeNames.size() == 1;
+
+		this.entityKey = entityKey;
```
```entityKey``` must be non-null?

sebersole replied:
Ofc

gbadner commented on hibernate-core/src/main/java/org/hibernate/persister/entity/AbstractEntityPersister.java:
```diff
@@ -4279,6 +4307,50 @@ public Object load(Serializable id, Object optionalObject, LockOptions lockOptio
+			);
+		}
+
+		throw new IllegalStateException(  );
```
Probably won't happen, but there should be a description of the exception.

gbadner commented on hibernate-core/src/main/java/org/hibernate/persister/entity/EntityPersister.java:
```diff
@@ -798,7 +814,11 @@ Object createProxy(Serializable id, SharedSessionContractImplementor session)
+
+	default BytecodeEnhancementMetadata getBytecodeEnhancementMetadata() {
+		return getInstrumentationMetadata();
+	}
```
```getBytecodeEnhancementMetadata()``` does not seem to be used anywhere. Should it be removed?

sebersole replied:
No.  Actually the other way around - that is the method we should use.  I got caught not finishing what I intended there.  I have changed that (and some other things) and will push soon

gbadner commented:
Supersedes https://github.com/hibernate/hibernate-orm/pull/2864

dreab8 commented on hibernate-core/src/main/java/org/hibernate/bytecode/enhance/spi/interceptor/EnhancementAsProxyLazinessInterceptor.java:
```diff
@@ -93,62 +160,94 @@ public Object forceInitialize(Object target, String attributeName) {
 
```
can we remove the` BytecodeLogger.LOGGER.tracef`? I think the `forceInitialize` method that is called after does it

gbadner commented on hibernate-core/src/main/java/org/hibernate/event/internal/DefaultLoadEventListener.java:
```diff
@@ -309,7 +309,7 @@ private Object proxyOrLoad(
 					entity,
 					Status.MANAGED,
-					ArrayHelper.filledArray( LazyPropertyInitializer.UNFETCHED_PROPERTY, Object.class, persister.getPropertyTypes().length ),
+					new Object[ persister.getPropertyTypes().length ],
```
Why is this change necessary? Is it possible that a loaded property will be mistaken for null?
